spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java

```diff
@@ -89,42 +94,91 @@
 	private ApplicationEventPublisher applicationEventPublisher;
 
 	/**
-	 * Use this constructor for a single url (although it may be overridden
-	 * if the server URI(s) are provided by the {@link MqttConnectOptions#getServerURIs()}
-	 * provided by the {@link MqttPahoClientFactory}).
+	 * Use this constructor for a single url (although it may be overridden if the server
+	 * URI(s) are provided by the {@link MqttConnectOptions#getServerURIs()} provided by
+	 * the {@link MqttPahoClientFactory}).
+	 * @deprecated in favor of
+	 * {@link #MqttPahoMessageDrivenChannelAdapter(List, String, MqttPahoClientFactory, String...)}
+	 * - set the url in the client factory connection options.
 	 * @param url the URL.
 	 * @param clientId The client id.
 	 * @param clientFactory The client factory.
 	 * @param topic The topic(s).
 	 */
+	@Deprecated
+	@SuppressWarnings("deprecation")
 	public MqttPahoMessageDrivenChannelAdapter(String url, String clientId, MqttPahoClientFactory clientFactory,
 			String... topic) {
 		super(url, clientId, topic);
 		this.clientFactory = clientFactory;
 	}
 
 	/**
-	 * Use this constructor if the server URI(s) are provided by the {@link MqttConnectOptions#getServerURIs()}
-	 * provided by the {@link MqttPahoClientFactory}.
+	 * Use this constructor with a list of server URIs, which will take precedence over
+	 * the factory's connection options' server URIs.
+	 * @param serverUris the server URIs.
+	 * @param clientId The client id.
+	 * @param clientFactory The client factory.
+	 * @param topic The topic(s).
+	 * @since 5.4
+	 */
+	public MqttPahoMessageDrivenChannelAdapter(List<String> serverUris, String clientId,
+			MqttPahoClientFactory clientFactory, String... topic) {
+		super(clientId, topic);
+		this.clientFactory = clientFactory;
+		this.clientFactory.setServerUris(serverUris);
```

Comment:
I wouldn't do this. It is mutation of an external shared factory.
What drove you to drop that `url`-based configuration?
I thought we just need to have a consistency on the read side...

Reply:
Oh, you're right - need to rethink.

The reason is if you want to connect to a cluster, the only way you can do it is via the connect options - here I am restoring the functionality, but enhanced to allow cluster per adapter with a single factory.